7-extract_plier_pathways.R

```diff
@@ -265,6 +272,21 @@ for (seed_index in 1:length(norm.train.files)) {
   norm_methods_else <- c("log", "npn", "qn", "qn-z", "tdm", "z",
                          "array_only", "seq_only")
 
+  # set random seeds to use inside %dopar% loop for each %seq and norm method
+  use_seeds_inside_dopar <- list()
+  for (ps in perc_seq) {
+    
+    if (ps %in% c("0", "100")) {
+      for (nm in norm_methods_if_0_100) {
+        use_seeds_inside_dopar[[ps]][[nm]] <- sample(1:1000, size = 1)
+      }
+    } else {
+      for (nm in norm_methods_else) {
+        use_seeds_inside_dopar[[ps]][[nm]] <- sample(1:1000, size = 1)
+      }
+    }
+  }
+  
```

Comment:
I wonder about the advantages of this approach, over something where you're finding the length of the flattened list (total number of seeds you'd need; `total_length_seeds` in a sec) and creating a vector of seeds with `sample(1:1000, size = total_length_seeds)`?  You could potentially then do something like `seed_index <- 1` before the loop starts and increment the index with `seed_index <- seed_index + 1`.

Things I might not be taking into account with this comment:

- There's something in the file naming that needs this structure.
- It's helpful for debugging in someway.
- The parallelization means that incrementing the index is not at all going to work how I might expect.

More generally though, if the command is exactly the same (`use_seeds_inside_dopar[[ps]][[nm]] <- sample(1:1000, size = 1)`), why is all this logic necessary? I'm less acquainted with the nitty gritty of this code base than you are at this point.

Reply:
Thank you 😸

I'm inclined to keep things as is, though I definitely see your point about extra logic. There are a few things going on that I think make an approach like this necessary (while less palatable that one would like...)

With `%dopar%` in the mix, I don't see a way to have a seed counter reliably access the same seed every time for a particular `ps` and `nm` combination. I think once we are inside the parallel loop we lose connection with the seed count updating in other loops -- we could increment in Loop A but Loop B doesn't know about that. Setting up a seed for each combination in advance and then accessing that later feels robust to that sort of ambiguity.

The other complication is we have some normalization methods used for 0% RNA-seq (and 100% RNA-seq) (for now, just LOG) while there are several normalization methods used at the intermediate %RNA-seqs (i.e. 50%). One way to simplify this and eliminate the logic steps would be to create a seed for every possible combination of %RNA-seq and normalization method, even though some (most) would never be accessed inside the parallel loops. The benefit I see to the existing logic structure is just that we clearly account for each combination in an out loud way.